File: src/Blokemon.CardGen/Domain/CardEntry.cs
using System.Collections.Immutable;

namespace Blokemon.CardGen.Domain;

/// <summary>An entry in a card's mechanics region.</summary>
public abstract record CardEntry
{
    private CardEntry(MechanicalId id, string name)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(name);
        Id = id;
        Name = name;
    }

    /// <summary>The mechanical identifier of the entry.</summary>
    public MechanicalId Id { get; }

    /// <summary>The printed name of the entry.</summary>
    public string Name { get; }

    /// <summary>Applies the function matching this entry.</summary>
    /// <typeparam name="TResult">The result of the applied function.</typeparam>
    /// <param name="ability">Applied to an Ability.</param>
    /// <param name="attack">Applied to an Attack.</param>
    /// <param name="rule">Applied to a Rule.</param>
    /// <returns>The result of the applied function.</returns>
    public abstract TResult Match<TResult>(
        Func<Ability, TResult> ability,
        Func<Attack, TResult> attack,
        Func<Rule, TResult> rule
    );

    /// <summary>An Ability entry.</summary>
    /// <param name="Id">The mechanical identifier of the entry.</param>
    /// <param name="Name">The printed name of the entry.</param>
    /// <param name="EffectText">The printed effect text.</param>
    public sealed record Ability(MechanicalId Id, string Name, string EffectText)
        : CardEntry(Id, Name)
    {
        /// <inheritdoc/>
        public override TResult Match<TResult>(
            Func<Ability, TResult> ability,
            Func<Attack, TResult> attack,
            Func<Rule, TResult> rule
        ) => ability(this);
    }

    /// <summary>An Attack entry.</summary>
    /// <param name="Id">The mechanical identifier of the entry.</param>
    /// <param name="Name">The printed name of the entry.</param>
    /// <param name="EnergyCost">The Energy printed in the gutter.</param>
    /// <param name="Damage">The printed Damage.</param>
    /// <param name="EffectText">The printed effect text, absent on a pure-Damage Attack.</param>
    public sealed record Attack(
        MechanicalId Id,
        string Name,
        ImmutableArray<BlokemonType> EnergyCost,
        Damage Damage,
        string? EffectText
    ) : CardEntry(Id, Name)
    {
        /// <inheritdoc/>
        public override TResult Match<TResult>(
            Func<Ability, TResult> ability,
            Func<Attack, TResult> attack,
            Func<Rule, TResult> rule
        ) => attack(this);
    }

    /// <summary>A Rule entry.</summary>
    /// <param name="Id">The mechanical identifier of the entry.</param>
    /// <param name="Name">The printed name of the entry.</param>
    /// <param name="EffectText">The printed effect text.</param>
    public sealed record Rule(MechanicalId Id, string Name, string EffectText) : CardEntry(Id, Name)
    {
        /// <inheritdoc/>
        public override TResult Match<TResult>(
            Func<Ability, TResult> ability,
            Func<Attack, TResult> attack,
            Func<Rule, TResult> rule
        ) => rule(this);
    }
}

File: src/Blokemon.CardGen/Domain/CardEntry.fs
namespace Blokemon.CardGen.Domain

open System
open System.Collections.Immutable

/// An entry in a card's mechanics region.
[<RequireQualifiedAccess>]
type CardEntry =
    /// A Blokemon Power entry.
    | PokemonPower of id: MechanicalId * name: string * effectText: string

    /// An Attack entry, whose effect text is absent on a pure-Damage Attack.
    | Attack of
        id: MechanicalId *
        name: string *
        energyCost: ImmutableArray<BlokemonType> *
        damage: Damage *
        effectText: string option

    /// A Rule entry.
    | Rule of id: MechanicalId * name: string * effectText: string

    /// The mechanical identifier of the entry.
    member this.Id =
        match this with
        | CardEntry.PokemonPower(id = id)
        | CardEntry.Attack(id = id)
        | CardEntry.Rule(id = id) -> id

    /// The printed name of the entry.
    member this.Name =
        match this with
        | CardEntry.PokemonPower(name = name)
        | CardEntry.Attack(name = name)
        | CardEntry.Rule(name = name) -> name

/// An entry in a card's mechanics region.
[<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
module CardEntry =

    /// Every entry prints a name, so a blank one is a defect in the authority rather than a card.
    let private printedName (name: string) =
        ArgumentException.ThrowIfNullOrWhiteSpace(name, nameof name)
        name

    /// A Blokemon Power entry.
    let pokemonPower id name effectText =
        CardEntry.PokemonPower(id, printedName name, effectText)

    /// An Attack entry.
    let attack id name energyCost damage effectText =
        CardEntry.Attack(id, printedName name, energyCost, damage, effectText)

    /// A Rule entry.
    let rule id name effectText =
        CardEntry.Rule(id, printedName name, effectText)
